Sound.cpp

AJenbo opened the pull request:
snd_stop_snd, sound_CreateSoundBuffer, sound_file_cleanup, snd_init, sound_DirectSoundCreate, sound_cleanup, music_stop, sound_disable_music, sound_get_or_set_music_volume and sound_get_or_set_sound_volume where cleaned up. About half of them where already bin exact.

squidcc commented on Source/sound.cpp:
```diff
@@ -261,82 +260,69 @@ TSnd *__fastcall sound_file_load(char *path)
+	DSB.dwFlags = DSBCAPS_CTRLVOLUME|DSBCAPS_CTRLPAN|DSBCAPS_STATIC;
+	int soundBuffer = sglpDS->CreateSoundBuffer(&DSB, &sound_file->DSB, NULL);
+
+	if ( soundBuffer )
```
"int soundbuffer" should probably be "HRESULT hr" and this comparison "if ( hr != ERROR_SUCCESS )"

squidcc commented on Source/sound.cpp:
```diff
@@ -71,13 +71,12 @@ void __fastcall snd_update(bool bStopAll)
-			v1->Stop();
+	if ( pSnd ) {
+		DSB = pSnd->DSB;
+		if ( DSB )
```
Could be one line? "if ( pSnd && pSnd->DSB )"

squidcc replied:
I don't see why not? It's done that way further down.

squidcc commented on Source/sound.cpp:
```diff
@@ -261,82 +260,69 @@ TSnd *__fastcall sound_file_load(char *path)
-			v1->DSB->Release();
-			v1->DSB = 0;
+			sound_file->DSB->Release();
+			sound_file->DSB = 0;
```
Another style issue: do we insist on NULL for pointers instead of 0? I think we should.

mewmew replied:
I agree, use NULL for NULL pointers.

squidcc commented on Source/sound.cpp:
```diff
@@ -261,82 +260,69 @@ TSnd *__fastcall sound_file_load(char *path)
 	gbMusicOn = sglMusicVolume > -1600;
+
 	if ( sound_DirectSoundCreate(NULL, &sglpDS, NULL) )
 		sglpDS = 0;
```
Another 0 vs. NULL here

squidcc commented on Source/sound.cpp:
```diff
@@ -261,82 +260,69 @@ TSnd *__fastcall sound_file_load(char *path)
 	SVidInitialize(sglpDS);
 	SFileDdaInitialize(sglpDS);
+
 	gbSndInited = sglpDS != 0;
```
0 vs. NULL again

squidcc commented on Source/sound.cpp:
```diff
@@ -261,82 +260,69 @@ TSnd *__fastcall sound_file_load(char *path)
 {
-	int *v2; // esi
-	//int v3; // eax
 	int v4; // ecx
```
Could name this "vol" just to tidy up

AJenbo replied:
Actually I baled on that one, it's definatly not bin exact so shoudn't have been committed.

squidcc commented on Source/sound.cpp:
```diff
@@ -382,44 +368,32 @@ void __fastcall sound_create_primary_buffer(int music_track)
-		v6 = GetLastError();
-		ErrDlg(IDD_DIALOG5, v6, "C:\\Src\\Diablo\\Source\\SOUND.CPP", 427);
+
+	FARPROC proc = GetProcAddress(hDsound_dll, "DirectSoundCreate");
```
proc should be properly declared to match the function definition of DirectSoundCreate and the cast below removed.

AJenbo replied:
I have done my best to fix this now, but I'm not sure if it's as should. Might it work to include dsound.h and then have access to function without this or would it not be bin exact?

squidcc commented on Source/sound.cpp:
```diff
@@ -382,44 +368,32 @@ void __fastcall sound_create_primary_buffer(int music_track)
 
 int __fastcall sound_DirectSoundCreate(GUID *guid, IDirectSound **DS, int always_null)
```
"int always_null" should be "LPUNKNOWN pUnkOuter".

squidcc commented on Source/sound.cpp:
```diff
@@ -382,44 +368,32 @@ void __fastcall sound_create_primary_buffer(int music_track)
+
+	if ( sglpDS ) {
 		sglpDS->Release();
 		sglpDS = 0;
```
0 vs. NULL

AJenbo replied:
One note about it is fine 😉

squidcc commented on Source/sound.cpp:
```diff
@@ -468,36 +441,32 @@ void __fastcall music_start(int nTrack)
-	}
-	else if ( sgnMusicTrack != 6 )
-	{
+	} else if ( sgnMusicTrack != 6 ) {
```
Looking at the _music_id enum, looks like there should be "TMUSIC_NONE = 6" instead of "NUM_MUSIC = 6" and it should be used here (and above in music_stop)

AJenbo replied:
Not sure about this one, the other values of that enum dosn't seam to add up here. But maybe sgnMusicTrack[NUM_MUSIC] would make sens

AJenbo commented:
Highlighted issues should now be addressed, please give it a second review, thanks.

squidcc commented:
Looks good but I still think TMUSIC_NONE should be used instead of 6, the other enums match up fine with the names in sgszMusicTracks and other places as well e.g. menu_music_track_id starts off as TMUSIC_INTRO.

AJenbo commented:
Since that enum comes from the PSX build I'm not much inclined to change it

AJenbo commented:
Also cleaned up snd_update, snd_playing and snd_play_snd. I added them to this PR as a second commit (for easier review) since they depend on my previous commit to this file.

AJenbo commented:
I'm using WinQuake as a reference for some of the clean up since it was made around the same time and also uses DSound:
https://github.com/id-Software/Quake/blob/bf4ac424ce754894ac8f1dae6a3981954bc9852d/QW/client/snd_win.c#L212

AJenbo commented:
Add a commit that clean up sound_dup_channel, sound_file_reload, sound_file_load and sound_create_primary_buffer. That's all of the file then.

AJenbo commented on Source/sound.cpp:
```diff
@@ -46,381 +46,312 @@ struct sound_cpp_init
+
+		DSBs[i]->Stop();
+		DSBs[i]->Release();
+		DSBs[i] = 0;
```
Thanks, fixed the issues pointed out in your latest comments.

mewmew commented on Source/sound.cpp:
```diff
@@ -46,381 +46,312 @@ struct sound_cpp_init
-	IDirectSoundBuffer *result; // eax
-	IDirectSoundBuffer **v2; // esi
+	if ( !gbDupSounds ) {
+		return 0;
```
`NULL`, also for other returns.

mewmew commented on Source/sound.cpp:
```diff
@@ -46,381 +46,312 @@ struct sound_cpp_init
-
-	v2 = path;
 	if ( !sglpDS )
 		return 0;
```
Likewise, NULL return

AJenbo replied:
Thanks, all 4 are now fixed.

AJenbo commented:
@mewmew hmm ok should be fixed here:
https://github.com/diasurgical/devilution/pull/371

galaxyhaxz commented:
This shouldn't be happening. Both `min/max` are defined in `windef.h` included by `windows.h`. In both MSVC 4.0+ and GCC compilers. So for some reason it seems to be excluding that file and/or macro.

AJenbo commented:
Odd both GCC and LLVM builds where failing, but all VC where passing.

squidcc commented:
GCC might exclude it for C++ to avoid conflicting with std::min/max.
